DROPPING COLUMNS AND CHANGING THE INDEX OF A DATAFRAME

In [1]:
#import required modules
import pandas as pd
import numpy as np

In [2]:
#import and create dataframe
df = pd.read_csv('Dataset/BL-Flickr-Images-Book.csv')
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [3]:
#drop unwanted column
to_drop = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
            'Issuance type',
           'Shelfmarks']
df.drop(to_drop, inplace = True, axis = 1)   
#inplace = true indicates change directly into object
#another way : df.drop(columns=to_drop, inplace=True)
# this way remove the columns by passing them to the columns paraemter directly

In [4]:
#check the dataframe again
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

In [5]:
#check the uniqueness of Idenifier
df['Identifier'].is_unique

True

In [6]:
#replace existing index with Identifier
df = df.set_index ('Identifier')
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

In [7]:
#to access each record in a straightforward way using loc
df.loc[206]

Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

In [8]:
#using iloc, -1 is last data
df.iloc[-1]

Place of Publication                                               London
Date of Publication                                               1834-43
Publisher                                                             NaN
Title                   Collectanea Topographica et Genealogica. [Firs...
Author                                                                NaN
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4160339, dtype: object

TIDYING UP FIELDS IN THE DATA
cleaning Date of Publication and Place of Publication

In [9]:
#check datatype of the columns
df.dtypes

Place of Publication    object
Date of Publication     object
Publisher               object
Title                   object
Author                  object
Flickr URL              object
dtype: object

Date of Publication should not in object

In [10]:
df.loc[1905:, 'Date of Publication'].head (10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

A particular book can have only one date of publication. Therefore, we need to do the following:

1. Remove the extra dates in square brackets, wherever present: 1879 [1878]
2. Convert date ranges to their “start date”, wherever present: 1860-63; 1839, 38-54
3. Completely remove the dates we are not certain about and replace them with NumPy’s NaN: [1897?]
4. Convert the string nan to NumPy’s NaN value

synthesize the pattern of date using single regular expression

regex = r'^(\d{4})'

its meant to find  any four digit at the beginning of a string.
\d represents any digit, {4} repeats this rule 4 times
 ^ matches the start of a string, () denotes a capturing group, which signals to Pandas that we want
to extract that part of the regex. (want ^ to avoid cases where [] starts off the string)

In [14]:
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand = False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [46]:
#convert dtype of date  to numeric (since only have year instead of full date)
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

In [42]:
df['Date of Publication'].head()

Identifier
206    1879.0
216    1868.0
218    1869.0
472    1851.0
480    1857.0
Name: Date of Publication, dtype: float64

In [16]:
df['Date of Publication'].isnull().sum()/len(df)

0.11717147339205986

This results in about one in every ten values being missing, which is a small price to pay for now being able to do computations on the remaining valid values

COMBINING str METHODS WITH NUMPY TO CLEAN COLUMNS

In [26]:
df['Place of Publication'].unique()

array(['London', 'London; Virtue & Yorston',
       'pp. 40. G. Bryan & Co: Oxford, 1898', ...,
       'pp. viii. 64. J. Debrett: London, 1789', 'G. Eld: London, 1608',
       'Newcastle upon Tyne'], dtype=object)

In [17]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

from some row, we can see there is unnecessary information especially at London and oxford.

In [18]:
#Another situation
df.loc[4157862]

Place of Publication                                  Newcastle-upon-Tyne
Date of Publication                                                  1867
Publisher                                                      T. Fordyce
Title                   Local Records; or, Historical Register of rema...
Author                      FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4157862, dtype: object

In [19]:
df.loc[4159587]

Place of Publication                                  Newcastle upon Tyne
Date of Publication                                                  1834
Publisher                                                Mackenzie & Dent
Title                   An historical, topographical and descriptive v...
Author                                              Mackenzie, E. (Eneas)
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4159587, dtype: object

there is a different in the existence of hyphens

In [20]:
#use str.contains() to get a Boolean mask. str.contains is used to test if pattern is contain within a string
pub = df['Place of Publication']
london = pub.str.contains ('London')
london [:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [27]:
oxford = pub.str.contains ('Oxford')

In [37]:
#combine them with np.where . np.where(condition,then,else)
df['Place of Publication'] = np.where (london, 'London' ,
                                       np.where (oxford, 'Oxford', 
                                                pub.str.replace('-', '  ')))
df['Place of Publication'].head()

Identifier
206    London
216    London
218    London
472    London
480    London
Name: Place of Publication, dtype: object

In [41]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

CLEANING THE ENTIRE DATASET USING THE applymap FUNCTION
using "university_towns.txt"
The applymap() method took each element from the DataFrame, passed it to the function, and the original value was replaced by the returned value

In [51]:
university_towns = []
with open ('Dataset/university_towns.txt') as file:
    for line in file:
        if '[edit]' in line : 
            # Remember this `state` until the next is found
            state = line
        else:
            # Otherwise, we have a city; keep `state` as last-seen
            university_towns.append((state,line))
university_towns[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [53]:
#wrap list above into dataframe
towns_df = pd.DataFrame(university_towns, columns = ['State','Region Name'])
towns_df.head()

State                                        Region Name
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

In [54]:
def get_citystate (item):
    if '(' in item :
        return item [:item.find('(')] #return item until find ( , (:)
    elif '[' in item :
        return item [:item.find('[')] ##return item until find [
    else :
        return item
#item is element of the data


In [55]:
towns_df = towns_df.applymap(get_citystate)
#panda's .applymap() only take 1 parameter

In [56]:
towns_df.head()

State    Region Name
0  Alabama        Auburn 
1  Alabama      Florence 
2  Alabama  Jacksonville 
3  Alabama    Livingston 
4  Alabama    Montevallo

RENAMING COLUMNS AND SKIPPING ROWS
using 'olympics.csv' dataset

In [57]:
olympics_df = pd.read_csv('Dataset/olympics.csv')
olympics_df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  ? Summer  01 !  02 !  03 !  Total  ? Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  ? Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

This data is messy.

Therefore, we need to do two things:

1. Skip one row and set the header as the first (0-indexed) row
2. Rename the columns

In [58]:
#skip row 0 and set the header at 1
olympics_df = pd.read_csv('Dataset/olympics.csv',header =1 )
olympics_df.head()

Unnamed: 0  ? Summer  01 !  02 !  03 !  Total  ? Winter  \
0        Afghanistan (AFG)        13     0     0     2      2         0   
1            Algeria (ALG)        12     5     2     8     15         3   
2          Argentina (ARG)        23    18    24    28     70        18   
3            Armenia (ARM)         5     1     2     9     12         6   
4  Australasia (ANZ) [ANZ]         2     3     4     5     12         0   

   01 !.1  02 !.1  03 !.1  Total.1  ? Games  01 !.2  02 !.2  03 !.2  \
0       0       0       0        0       13       0       0       2   
1       0       0       0        0       15       5       2       8   
2       0       0       0        0       41      18      24      28   
3       0       0       0        0       11       1       2       9   
4       0       0       0        0        2       3       4       5   

   Combined total  
0               2  
1              15  
2              70  
3              12  
4              12

In [63]:
#rename the column
new_names = {'Unnamed: 0':'Country','? Summer':'Summer Olympic',
             '01 !':'Gold','02 !':'Silver','03 !':'Bronze',
             '? Winter':'Winter Olympics', '01 !.1':'Gold.1',
            '02 !.1':'Silver.1','03 !.1':'Bronze.1','? Games':'Total Games',
             '01 !.2':'Total Gold','02 !.2':'Total Silver','03 !.2':'Total Bronze'}

In [64]:
olympics_df.rename(columns = new_names, inplace = True)

In [65]:
olympics_df.head()

Country  Summer Olympicd  Gold  Silver  Bronze  Total  \
0        Afghanistan (AFG)               13     0       0       2      2   
1            Algeria (ALG)               12     5       2       8     15   
2          Argentina (ARG)               23    18      24      28     70   
3            Armenia (ARM)                5     1       2       9     12   
4  Australasia (ANZ) [ANZ]                2     3       4       5     12   

   Winter Olympics  Gold.1  Silver.1  Bronze.1  Total.1  Total Games  \
0                0       0         0         0        0           13   
1                3       0         0         0        0           15   
2               18       0         0         0        0           41   
3                6       0         0         0        0           11   
4                0       0         0         0        0            2   

   Total Gold  Total Silver  Total Bronze  Combined total  
0           0             0             2               2  
1           5             2             8              15  
2          18            24            28              70  
3           1             2             9              12  
4           3             4             5              12

In [ ]:
#to see overview  data frame for row that have missing data only
null_data = train[train.isnull().any(axis=1)]
null_data

In [ ]:
#to print certain row that their column have certain value
train.loc[train['employee_id'] == 71177]

In [ ]:
#create new data frame with new 
promoted = train[train.is_promoted == 1]
promoted